src/marks/density.js

```diff
@@ -11,44 +12,113 @@ const defaults = {
 };
 
 export class Density extends Mark {
-  constructor(data, options = {}) {
-    const {x, y, bandwidth = 20, thresholds = 20} = options;
+  constructor(data, {x, y, z, weight, stroke, fill, bandwidth = 20, thresholds = 20, ...options} = {}) {
+    let f, s;
+    if (fill === "density") { fill = undefined; f = true; }
+    if (stroke === "density") { stroke = undefined; s = true; }
     super(
       data,
       [
         {name: "x", value: x, scale: "x", optional: true},
-        {name: "y", value: y, scale: "y", optional: true}
+        {name: "y", value: y, scale: "y", optional: true},
+        {name: "weight", value: weight, optional: true},
+        {name: "z", value: maybeZ({z, fill, stroke}), optional: true}
       ],
-      options,
+      densityInitializer({...options, fill, stroke}, +bandwidth, +thresholds, f, s),
       defaults
     );
-    this.bandwidth = +bandwidth;
-    this.thresholds = +thresholds;
+    this.z = z;
+    this.path = geoPath();
+  }
+  filter(index) {
+    return index;
   }
   render(index, scales, channels, dimensions) {
-    const {x: X, y: Y} = channels;
-    const {bandwidth, thresholds} = this;
-    const [cx, cy] = applyFrameAnchor(this, dimensions);
-    const {width, height} = dimensions;
+    const {contours} = channels;
+    const {path} = this;
     return create("svg:g")
         .call(applyIndirectStyles, this, scales, dimensions)
         .call(applyTransform, this, scales)
-        .call(g => g.selectAll("path")
-          .data(contourDensity()
-              .x(X ? i => X[i] : constant(cx))
-              .y(Y ? i => Y[i] : constant(cy))
-              .size([width, height])
-              .bandwidth(bandwidth)
-              .thresholds(thresholds)
-            (index))
+        .call(g => g.selectAll()
+          .data(Array.from(index, i => [i]))
           .enter()
           .append("path")
-            .attr("d", geoPath()))
-      .node();
+            .call(applyDirectStyles, this)
+            .call(applyGroupedChannelStyles, this, channels)
+            .attr("d", ([i]) => path(contours[i])))
+        .node();
   }
 }
 
 export function density(data, {x, y, ...options} = {}) {
   ([x, y] = maybeTuple(x, y));
   return new Density(data, {...options, x, y});
 }
+
+function densityInitializer(options, bandwidth, thresholds, f, s) {
+  return initializer(options, function(data, facets, channels, scales, dimensions) {
+    const X = channels.x?.scale ? valueof(channels.x.value, scales[channels.x.scale]) : channels.x?.value;
+    const Y = channels.y?.scale ? valueof(channels.y.value, scales[channels.y.scale]) : channels.y?.value;
+    const W = channels.weight?.value;
+    const Z = channels.z?.value;
+    const {z} = this;
+    const [cx, cy] = applyFrameAnchor(options, dimensions);
```

Comment:
`applyFrameAnchor` takes `this` as the first argument, not `options`. (I’ll fix this.)